src/app/api/auth/lib/nextauth-options.ts

```diff
@@ -88,39 +150,78 @@ export const nextAuthOptions: AuthOptions = {
       return false
     },
     session: async ({ session, token }) => {
-      // TODO: Need to figure out how to do this more efficiently
-      // Check if the user has a valid accessToken
-      // const validSession = await verifySession(token?.accessToken as string);
+      if (token) {
+        session.user.name = token.name as string
+        session.user.image = token.image as string
+        session.user.email = token.email as string
+        session.user.accessToken = token.accessToken as string
+        session.expires = new Date(
+          token.accessTokenExpires as number,
+        ).toISOString()
+        session.error = token.error
+      }
 
-      // if (!validSession) {
-      //   return undefined as any;
-      // }
+      return session
+    },
+    jwt: async ({ token, user, account }) => {
+      // Initial sign in
+      if (user && account) {
+        authLogger.debug('Initial sign in')
+        console.log('account', account)
+        token = {
+          accessToken: account.access_token as string,
+          // Access token expiration is provided as number in seconds since epoch (value is always 8 hours)
+          accessTokenExpires: (account.expires_at as number) * 1000,
+          refreshToken: account.refresh_token as string,
+          // Refresh token expiration is provided as number of seconds until expiration (value is always 6 months)
+          refreshTokenExpires:
+            Date.now() + (account.refresh_token_expires_in as number) * 1000,
+          ...user,
+        }
+
+        return token
+      }
 
-      // This is fine when the session is invalid
-      if (!token) {
-        return undefined as any
+      // Return previous token if the access token has not expired yet
+      if (
+        token.accessTokenExpires &&
```

Comment:
is `token.accessTokenExpires` always true? If so, can we remove?
Otherwise, if its not always true, how are we handling that case. If I'm reading this correctly it looks like if `token.accessTokenExpires` is false, it falls through these logic checks and assumes the token is expired, trying to refresh. Is that desired?

Reply:
This makes it seem like it's always true: `// Access token expiration is provided as number in seconds since epoch (value is always 8 hours)`

Reply:
Yes this can be removed, a JWT cannot be generated without that because of the above `if`

```typescript
token = {
          accessToken: account.access_token as string,
          // Access token expiration is provided as number in seconds since epoch (value is always 8 hours)
          accessTokenExpires: secondsToMilliseconds(
            account.expires_at as number,
          ),
          refreshToken: account.refresh_token as string,
          // Refresh token expiration is provided as number of seconds until expiration (value is always 6 months)
          refreshTokenExpires:
            Date.now() +
            secondsToMilliseconds(account.refresh_token_expires_in as number),
          ...user,
        }

        return token
```

Reply:
(Though I wonder if previously signed JWTs have this field 🤔)